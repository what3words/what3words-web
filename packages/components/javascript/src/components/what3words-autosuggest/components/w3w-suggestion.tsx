import type { FunctionalComponent } from "@stencil/core";
import type { JSXBase } from "@stencil/core/internal";
import { DEFAULTS } from "@javascript-components/lib/constants";
import { t } from "@javascript-components/lib/translation";
import { h } from "@stencil/core";

import type { AutosuggestOption } from "../../what3words-autosuggest/domain";

const c = "what3words-autosuggest";

type W3wSuggestionProps = JSXBase.HTMLAttributes<HTMLDivElement> & {
  opt: AutosuggestOption;
};

export const W3wSuggestion: FunctionalComponent<W3wSuggestionProps> = ({
  opt,
  ...props
}) => {
  const {
    country,
    distanceToFocusKm,
    nearestPlace = "",
    words,
    language,
  } = opt;

  return (
    <div {...props}>
      <div class={`${c}-address`} data-testid="address">
        <what3words-symbol
          size={DEFAULTS.iconSize}
          color={DEFAULTS.iconColor}
        />
        <div class={`${c}-words`} data-testid="words">
          {words}
        </div>
      </div>
      <div class={`${c}-nearest-place`} data-testid="nearest-place">
        {country.toLowerCase() === "zz" ? (
          <div
            class={`${c}-flag ${c}-flag-${country.toLowerCase()}`}
            data-testid="flag"
          />
        ) : null}
        <div class={`${c}-nearest-place-text`} data-testid="nearest-place-text">
          {nearestPlace && (
            <div>
              {t("nearest_place", {
                param: nearestPlace,
                language,
                strict: false,
              })}
            </div>
          )}
          {distanceToFocusKm ? (
            <div
              class={`${c}-nearest-place-distance`}
              data-testid="nearest-place-distance"
            >
              {distanceToFocusKm} km
            </div>
          ) : null}
        </div>
      </div>
    </div>
  );
};
