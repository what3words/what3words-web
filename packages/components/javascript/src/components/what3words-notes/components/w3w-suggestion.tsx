import type { FunctionalComponent } from "@stencil/core";
import type { JSXBase } from "@stencil/core/internal";
import { DEFAULTS } from "@javascript-components/lib/constants";
import { h } from "@stencil/core";

import type { AutosuggestOption } from "../../what3words-autosuggest/domain";
import SeaIcon from "../assets/sea.svg";

type W3wSuggestionProps = JSXBase.HTMLAttributes<HTMLLIElement> & {
  tag: string;
  opt: AutosuggestOption;
};

export const W3wSuggestion: FunctionalComponent<W3wSuggestionProps> = ({
  tag,
  opt,
  ...props
}) => {
  return (
    <li {...props}>
      <div class={`${tag}-address`} data-testid="address">
        <what3words-symbol size={14} color={DEFAULTS.iconColor} />
        <div class={`${tag}-words`} data-testid="words">
          {opt.words}
        </div>
      </div>
      <div class={`${tag}-nearest-place`} data-testid="nearest-place">
        <div
          class={`${tag}-nearest-place-text`}
          data-testid="nearest-place-text"
        >
          {opt.nearestPlace ? (
            opt.nearestPlace
          ) : (
            <img class={`${tag}-sea-icon`} src={SeaIcon} alt="Ocean" />
          )}
        </div>
        {opt.distanceToFocusKm && !isNaN(opt.distanceToFocusKm) && (
          <div
            class={`${tag}-nearest-place-distance`}
            data-testid="nearest-place-distance"
          >
            {opt.distanceToFocusKm}km
          </div>
        )}
      </div>
    </li>
  );
};
