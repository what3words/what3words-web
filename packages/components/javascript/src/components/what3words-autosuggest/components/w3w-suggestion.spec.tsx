import { h } from "@stencil/core";
import { newSpecPage } from "@stencil/core/testing";

import type { AutosuggestOption } from "../domain";
import { W3wSuggestion } from "./w3w-suggestion";

const defaultOption: AutosuggestOption = {
  country: "GB",
  distanceToFocusKm: 1,
  nearestPlace: "London",
  words: "apple.banana.cherry",
  language: "en",
  rank: 1,
};

describe("Render w3w-suggestion", () => {
  it("render correct suggestion", async () => {
    const page = await newSpecPage({
      components: [],
      template: () => <W3wSuggestion opt={defaultOption}></W3wSuggestion>,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should not now show nearest place when not provided", async () => {
    const option: AutosuggestOption = {
      ...defaultOption,
      nearestPlace: "",
    };

    const page = await newSpecPage({
      components: [],
      template: () => <W3wSuggestion opt={option}></W3wSuggestion>,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should show sea flag when country is ZZ", async () => {
    const option: AutosuggestOption = {
      ...defaultOption,
      country: "ZZ",
    };

    const page = await newSpecPage({
      components: [],
      template: () => <W3wSuggestion opt={option}></W3wSuggestion>,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should not show distance when not provided", async () => {
    const option: AutosuggestOption = {
      ...defaultOption,
      distanceToFocusKm: undefined,
    };

    const page = await newSpecPage({
      components: [],
      template: () => <W3wSuggestion opt={option}></W3wSuggestion>,
    });

    expect(page.root).toMatchSnapshot();
  });
});
