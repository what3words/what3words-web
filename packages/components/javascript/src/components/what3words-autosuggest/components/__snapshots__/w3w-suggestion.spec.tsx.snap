// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render w3w-suggestion render correct suggestion 1`] = `
<div>
  <div class="what3words-autosuggest-address" data-testid="address">
    <what3words-symbol color="#e11f26" size="17"></what3words-symbol>
    <div class="what3words-autosuggest-words" data-testid="words">
      apple.banana.cherry
    </div>
  </div>
  <div class="what3words-autosuggest-nearest-place" data-testid="nearest-place">
    <div class="what3words-autosuggest-nearest-place-text" data-testid="nearest-place-text">
      <div>
        near London
      </div>
      <div class="what3words-autosuggest-nearest-place-distance" data-testid="nearest-place-distance">
        1 km
      </div>
    </div>
  </div>
</div>
`;

exports[`Render w3w-suggestion should not now show nearest place when not provided 1`] = `
<div>
  <div class="what3words-autosuggest-address" data-testid="address">
    <what3words-symbol color="#e11f26" size="17"></what3words-symbol>
    <div class="what3words-autosuggest-words" data-testid="words">
      apple.banana.cherry
    </div>
  </div>
  <div class="what3words-autosuggest-nearest-place" data-testid="nearest-place">
    <div class="what3words-autosuggest-nearest-place-text" data-testid="nearest-place-text">
      <div class="what3words-autosuggest-nearest-place-distance" data-testid="nearest-place-distance">
        1 km
      </div>
    </div>
  </div>
</div>
`;

exports[`Render w3w-suggestion should not show distance when not provided 1`] = `
<div>
  <div class="what3words-autosuggest-address" data-testid="address">
    <what3words-symbol color="#e11f26" size="17"></what3words-symbol>
    <div class="what3words-autosuggest-words" data-testid="words">
      apple.banana.cherry
    </div>
  </div>
  <div class="what3words-autosuggest-nearest-place" data-testid="nearest-place">
    <div class="what3words-autosuggest-nearest-place-text" data-testid="nearest-place-text">
      <div>
        near London
      </div>
    </div>
  </div>
</div>
`;

exports[`Render w3w-suggestion should show sea flag when country is ZZ 1`] = `
<div>
  <div class="what3words-autosuggest-address" data-testid="address">
    <what3words-symbol color="#e11f26" size="17"></what3words-symbol>
    <div class="what3words-autosuggest-words" data-testid="words">
      apple.banana.cherry
    </div>
  </div>
  <div class="what3words-autosuggest-nearest-place" data-testid="nearest-place">
    <div class="what3words-autosuggest-flag what3words-autosuggest-flag-zz" data-testid="flag"></div>
    <div class="what3words-autosuggest-nearest-place-text" data-testid="nearest-place-text">
      <div>
        near London
      </div>
      <div class="what3words-autosuggest-nearest-place-distance" data-testid="nearest-place-distance">
        1 km
      </div>
    </div>
  </div>
</div>
`;
